srv/salt/_modules/osd.py

```diff
@@ -232,6 +233,214 @@ def tree_from_any():
     return _tree(**kwargs)
 
 
+def _get_parent_device_name(device_name):
+    """
+    Get the parent device name.
+
+      * sdb1 => sdb
+      * dm-0 => sdg
+      * vda => vda
+
+    :param device_name: The device name, e.g. vdd1, dm-0
+    :return: Returns the parent device name.
+    :rtype: str
+    """
+    # Check if given name is a partition, e.g. 'vdd1', or 'sdb2'.
+    parts = re.match(r'^([a-zA-Z]+)(\d+)$', device_name)
+    if parts is not None:
+        return parts.group(1)
+    # Check if it is a device mapper, e.g. 'dm-1'.
+    if re.match(r'^dm-(\d+)$', device_name):
+        # Get the slaves of the device.
+        dm_slaves = [
+            os.path.basename(slave) for slave in glob.glob(
+                '/sys/block/{}/slaves/*'.format(device_name))
+        ]
+        # Use the first and mostly single entry.
+        # /sys/devices/pci0000:00/0000:00:06.0/virtio3/block/vdd/vdd1 => vdd1
+        return _get_parent_device_name(dm_slaves[0])
+    return device_name
+
+
+def metadata():
+    """
+    Get the OSD metadata.
+    :return: Returns the OSD metadata.
+    :rtype: list
+    """
+    cluster = rados.Rados(conffile='/etc/ceph/ceph.conf')
+    cluster.connect()
+    cmd = json.dumps({'prefix': 'osd metadata', 'format': 'json'})
+    _, output, _ = cluster.mon_command(cmd, b'', timeout=6)
+    return json.loads(output)
+
+
+def details(osd_metadata=None):
+    """
+    Get a list of OSD's details. This includes information about the used
+    devices for the primary/data, journal/wal and database parts.
+
+    .. code-block:: json
+        [{
+            "cluster": "ceph",
+            "id": 5,
+            "name": "osd.5",
+            "hostname": "data1",
+            "type": "bluestore",
+            "primary": {
+                "rotational": true,
+                "device_name": "vdg",
+                "device_file": "/dev/vdg"
+            },
+            "journal": {
+                "rotational": null,
+                "device_name": null,
+                "device_file": null
+            },
+            "db": {
+                "rotational": null,
+                "device_name": null,
+                "device_file": null
+            }
+        }]
+
+    :param osd_metadata: The OSD metadata.
+    :type osd_metadata: list
+    :return: A list of OSD's and their used devices.
+    :rtype: list
+    """
+    result = []
+
+    if not osd_metadata:
+        return result
+
+    # Get a list of OSD ID's on this host.
+    osd_ids = list_()
+    if not osd_ids:
+        return result
```

Comment:
A logging message would be nice here:

"No OSDs found on host {self}. Aborting"

Reply:
Done